design-proposals/kubevirt_ephemeral_csi_driver.md

```diff
@@ -0,0 +1,79 @@
+
+If we use a csi ephemeral volume controlled by a kubevirt specific csi driver, we can control the behavior in that scenario and unmount any mounted volumes inside the csi ephemeral volume before clearing it.
+
+## Goals
+
+* Design and implement a kubevirt specific csi driver that only supports **ephemeral volumes**.
+* Use csi ephemeral volumes to fully control what happens when virt-launcher pods terminate with regards to volume hotplug.
+
+## Non Goals
+
+Implement a full fledged CSI driver that also supports persistent volumes. We have other projects like the [hostpath csi driver](https://github.com/kubevirt/hostpath-provisioner) for that.
+
+
+## User Stories
+
+* As a KubeVirt user I want to be guaranteed that if I have hotplugged volumes in my VM, and the virt-launcher pod terminates unexpectedly, I will not lose any data on my volumes.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+The idea is to implement just the ephemeral parts of the csi spec. Examples of drivers that do this are [csi-driver-image-populate](https://github.com/kubernetes-csi/csi-driver-image-populator) and [secrets-store-csi-driver](https://github.com/kubernetes-sigs/secrets-store-csi-driver).
+
+To support just the ephemeral portions, we need to implement the appropriate parts of the csi driver, and start the [node-driver-registrar](https://github.com/kubernetes-csi/node-driver-registrar) so we can register the driver with kubernetes. The csi driver will have to run on each node, so will require a daemonset. We could make an additional daemonset, or add additional containers to the virt-handler daemonset. The daemonset will require privileged permissions due to the nature of gRPC socket used to communicate between the different CSI components.
+
+In addition we will have to create the CSI driver object as shown below (I used example naming, we can pick something different if we want)
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: CSIDriver
+metadata:
+  name: kubevirt.csi.k8s.io
+  labels:
+    app.kubernetes.io/instance: kubevirt.csi.k8s.io
+    app.kubernetes.io/part-of: kubevirt.io
+    app.kubernetes.io/name: kubevirt.csi.k8s.io
+    app.kubernetes.io/component: csi-driver
+spec:
+  # Supports only ephemeral inline volumes.
+  volumeLifecycleModes:
+  - Ephemeral
+```
+
+One implementation detail to work out, is what to use where to create the volumes. The natural initial implemention would be a directory on the node somewhere, maybe the same directory as where other kubevirt specific configuration is stored?
```

Comment:
maybe a directory nested in the /var/run directory since it's not data that needs to persist node restarts

Reply:
Right, makes sense, and I think we already have a kubevirt specific directory in there.